File: hu/codes/javascript/chapter_computational_complexity/iteration.js
/**
 * File: iteration.js
 * Created Time: 2023-08-28
 * Author: Gaofer Chou (gaofer-chou@qq.com)
 */

/* for ciklus */
function forLoop(n) {
    let res = 0;
    // Összegzés: 1, 2, ..., n-1, n
    for (let i = 1; i <= n; i++) {
        res += i;
    }
    return res;
}

/* while ciklus */
function whileLoop(n) {
    let res = 0;
    let i = 1; // Feltételváltozó inicializálása
    // Összegzés: 1, 2, ..., n-1, n
    while (i <= n) {
        res += i;
        i++; // Feltételváltozó frissítése
    }
    return res;
}

/* while ciklus (két frissítéssel) */
function whileLoopII(n) {
    let res = 0;
    let i = 1; // Feltételváltozó inicializálása
    // Összegzés: 1, 4, 10, ...
    while (i <= n) {
        res += i;
        // Feltételváltozó frissítése
        i++;
        i *= 2;
    }
    return res;
}

/* Egymásba ágyazott for ciklus */
function nestedForLoop(n) {
    let res = '';
    // i ciklus: 1, 2, ..., n-1, n
    for (let i = 1; i <= n; i++) {
        // j ciklus: 1, 2, ..., n-1, n
        for (let j = 1; j <= n; j++) {
            res += `(${i}, ${j}), `;
        }
    }
    return res;
}

/* Tesztkód */
const n = 5;
let res;

res = forLoop(n);
console.log(`For loop sum result res = ${res}`);

res = whileLoop(n);
console.log(`While loop sum result res = ${res}`);

res = whileLoopII(n);
console.log(`While loop (two updates) sum result res = ${res}`);

const resStr = nestedForLoop(n);
console.log(`Nested for loop traversal result ${resStr}`);

File: hu/codes/java/chapter_computational_complexity/iteration.java
/**
 * File: iteration.java
 * Created Time: 2023-08-24
 * Author: krahets (krahets@163.com)
 */

package chapter_computational_complexity;

public class iteration {
    /* for ciklus */
    static int forLoop(int n) {
        int res = 0;
        // 1, 2, ..., n-1, n összegzése
        for (int i = 1; i <= n; i++) {
            res += i;
        }
        return res;
    }

    /* while ciklus */
    static int whileLoop(int n) {
        int res = 0;
        int i = 1; // Feltételváltozó inicializálása
        // 1, 2, ..., n-1, n összegzése
        while (i <= n) {
            res += i;
            i++; // Feltételváltozó frissítése
        }
        return res;
    }

    /* while ciklus (két frissítéssel) */
    static int whileLoopII(int n) {
        int res = 0;
        int i = 1; // Feltételváltozó inicializálása
        // 1, 4, 10, ... összegzése
        while (i <= n) {
            res += i;
            // Feltételváltozó frissítése
            i++;
            i *= 2;
        }
        return res;
    }

    /* Egymásba ágyazott for ciklus */
    static String nestedForLoop(int n) {
        StringBuilder res = new StringBuilder();
        // i = 1, 2, ..., n-1, n ciklus
        for (int i = 1; i <= n; i++) {
            // j = 1, 2, ..., n-1, n ciklus
            for (int j = 1; j <= n; j++) {
                res.append("(" + i + ", " + j + "), ");
            }
        }
        return res.toString();
    }

    /* Főprogram */
    public static void main(String[] args) {
        int n = 5;
        int res;

        res = forLoop(n);
        System.out.println("\nfor loop sum result res = " + res);

        res = whileLoop(n);
        System.out.println("\nwhile loop sum result res = " + res);

        res = whileLoopII(n);
        System.out.println("\nwhile loop (two updates) sum result res = " + res);

        String resStr = nestedForLoop(n);
        System.out.println("\nDouble for loop traversal result " + resStr);
    }
}
